### ISSUE TITLE
Keyboard game menu navigation is broken

### ISSUE THREAD

**Issue by attet** [widelands/widelands#5407]
First, the expected behavior from Widelands 1.0; when I navigate to a menu item and press space, it changes from this:

> ![1 0_pre](https://user-images.githubusercontent.com/1911416/167321076-4003459a-d025-4a4d-b013-17602700092e.png)

To this:

> ![1 0_after](https://user-images.githubusercontent.com/1911416/167321073-39efa5e4-0730-47fb-a70a-6ff03e58a6a1.png)

The text changed from 'Show statistics' to 'Hide statistics', and this setting is reflected in game. All is well and good.

Now compare this with the current master. Similarly we start from:
>  ![1 1_pre](https://user-images.githubusercontent.com/1911416/167321079-02ce4386-5b86-4479-8bef-362ca7489cef.png)

By pressing space we get to this:
>  ![1 1_after1](https://user-images.githubusercontent.com/1911416/167321077-45191bbd-b369-4d9d-a614-c933ec0d8d97.png)

Notice that:

- the menu oddly gets quite a bit wider (Some underlying text width gets recalculated?)
- the selection jumps to the topmost item
- most importantly the desired census status change is not triggered (no change in menu text nor in game).

Catastrophically, pressing space once more brings us to this weird state where the menu is there but all content is gone:
>  ![1 1_after2](https://user-images.githubusercontent.com/1911416/167321078-f3a1b84c-5da3-4ad8-a8e1-8116b3f58955.png)

**Version:**
 - OS: Windows 11
 - Widelands Version:  1.1~git25752[993b193@master] (Release)

**Comment by attet** [comment]
Just noticed that the overall main menu suffers from similar problems. E.g. in 'Single Player...' menu the space press makes Campaigns and Tutorials options disappear (and the whole menu is gone on second press):
![mainmenu1](https://user-images.githubusercontent.com/1911416/167323730-4cd256a0-9fe6-4d8e-a0ea-483c176e2f6e.png)
![mainmenu2](https://user-images.githubusercontent.com/1911416/167323732-c5d6276d-aa75-4da8-8d80-e32abf6fde27.png)

**Comment by Noordfrees** [comment]
The reason for this is that pressing character keys now filters the dropdown for the text you enter. Pressing Space once filters for entries containing a whitespace, pressing it again filters for strings containing two whitespaces in a row.

On the one hand, this is consistent with the new filtering. On the other hand I also find myself occasionally trying to select an element with Space instead of Enter. Should we restore the Space key behaviour to selecting an element instead of being interpreted as a character?
@jmoerschbach what do you think?

**Comment by attet** [comment]
I see. I found this behavior extremely confusing (not knowing about said filtering feature), especially as you can't actually open the menus (only the individual items) with enter. Juggling between space and enter in this way feels weird.

**Comment by matthiakl** [comment]
> On the one hand, this is consistent with the new filtering. 

No, it is not consistent: It only starts filtering on space or a key without a shortcut assigned. When the menu is open and you press `C` it toggles the census instead of filtering for c. 
Also it is now confusing, that you have to open the menu from the toolbar with space, but select an entry with enter (opening with enter does not work).

**Comment by tothxa** [comment]
Since all hotkeys have their function now even during filtering, I think `Space` should have it too. But ultimately this should be decoupled somehow. Some ideas:
- disable filtering in menus
- use a hotkey to start filtering, and disable hotkeys and `Space` in filter mode
- implement the requested but postponed visual feedback for filtering (this is important in any case)

**Comment by Noordfrees** [comment]
> Since all hotkeys have their function now even during filtering, I think `Space` should have it too.

+1

> * disable filtering in menus

That raises the question of which dropdowns are menus and which aren't. The main menu dropdowns wouldn't need filtering, but what about the editor tool menu? It's also a menu, but big enough for filtering to be useful. I think all dropdowns should remain filterable for consistency.

> * use a hotkey to start filtering, and disable hotkeys and `Space` in filter mode

-1 since the use of hotkeys is usually more important than filtering.
Or do you mean a *new* hotkey that puts the dropdown into "filter mode", while normal character presses no longer do that? That would work but it would also make the filtering feature less useful and harder to discover.

> * implement the requested but postponed visual feedback for filtering (this is important in any case)

+1
If someone wants to implement this soonish we can have it in v1.1

**Comment by tothxa** [comment]
> > * disable filtering in menus
> 
> That raises the question of which dropdowns are menus and which aren't.

In the worst case an extra argument could be added to dropdowns whether they need filtering. (default on, disable in main menus)

> what about the editor tool menu? It's also a menu, but big enough for filtering to be useful.

It's 11 entries, so I don't think it needs much filtering. I already said in https://github.com/widelands/widelands/pull/5365#issuecomment-1120274142 that they all need hotkeys though. (I guess I should just create a PR for this... :)

> > * use a hotkey to start filtering, and disable hotkeys and `Space` in filter mode
> 
> do you mean a _new_ hotkey that puts the dropdown into "filter mode", while normal character presses no longer do that?

Yes.

> That would work but it would also make the filtering feature less useful and harder to discover.

Unfortunately yes too.

**Comment by jmoerschbach** [comment]
sry, I haven't been around for a while...
Hope to have time again for widelands very soonish :)
I will have a look into why hotkey "space" is not triggered but instead directly affects the filter.
Selecting an item with "space" gets complicated as soon as "space" is also assigned as shortcut to another item... (imho in the end that means "space" cannot be assigned as hotkey if we want to select items via "space")
Highlighting the current search term is not yet forgotten and on my list (but a single whitespace might be tricky to highlight...)

**Comment by tothxa** [comment]
> but a single whitespace might be tricky to highlight...

You can change the background colour too, as already done for selection highlighting. That's visible for space as well.

**Comment by jmoerschbach** [comment]
hotkey "space" does work now and the currently active filter is highlighted as well -> closing this issue
